csharp/rust/src/ffi.rs

```diff
@@ -516,3 +534,54 @@ pub(crate) unsafe fn create_cmd(ptr: *const CmdInfo) -> Result<Cmd, String> {
     }
     Ok(cmd)
 }
+
+/// Convert [`BatchInfo`] to a [`Pipeline`].
+///
+/// # Safety
+/// * `ptr` must be able to be safely casted to a valid [`BatchInfo`].
+/// * `cmds` in a referred [`BatchInfo`] structure must not be `null`.
+/// * `cmds` in a referred [`BatchInfo`] structure must point to `cmd_count` consecutive [`CmdInfo`] pointers.
+///   They must be able to be safely casted to a valid to a slice of the corresponding type via [`from_raw_parts`]. See the safety documentation of [`from_raw_parts`].
+/// * Every pointer stored in `cmds` must not be `null` and must point to a valid [`CmdInfo`] structure.
+/// * All data in referred [`CmdInfo`] structure(s) should be valid. See the safety documentation of [`create_cmd`].
+pub(crate) unsafe fn create_pipeline(ptr: *const BatchInfo) -> Result<Pipeline, String> {
+    let cmd_pointers = unsafe { from_raw_parts((*ptr).cmds, (*ptr).cmd_count) };
+    let mut pipeline = Pipeline::with_capacity((*ptr).cmd_count);
+    for (i, cmd_ptr) in cmd_pointers.iter().enumerate() {
+        match unsafe { create_cmd(*cmd_ptr) } {
+            Ok(cmd) => pipeline.add_command(cmd),
+            Err(err) => return Err(format!("Coudln't create {:?}'th command: {:?}", i, err)),
+        };
+    }
+    if (*ptr).is_atomic {
+        pipeline.atomic();
+    }
+
+    Ok(pipeline)
+}
+
+/// Convert [`BatchOptionsInfo`] to a tuple of corresponding values.
+///
+/// # Safety
+/// * `ptr` could be `null`, but if it is not `null`, it must be a valid pointer to a [`BatchOptionsInfo`] struct.
+/// * `route_info` in dereferenced [`BatchOptionsInfo`] struct must contain a [`RouteInfo`] pointer.
+///   See description of [`RouteInfo`] and the safety documentation of [`create_route`].
+pub(crate) unsafe fn get_pipeline_options(
+    ptr: *const BatchOptionsInfo,
+) -> (Option<RoutingInfo>, Option<u32>, PipelineRetryStrategy) {
+    if ptr.is_null() {
+        return (None, None, PipelineRetryStrategy::new(false, false));
+    }
+    let timeout = if (*ptr).has_timeout {
+        Some((*ptr).timeout)
+    } else {
+        None
+    };
+    let route = unsafe { create_route((*ptr).route_info, None) };
+
+    (
+        route,
+        timeout,
+        PipelineRetryStrategy::new((*ptr).retry_server_error, (*ptr).retry_connection_error),
```

Comment:
I almost missed that there were pointer dereferences happening here. There's a bunch of places here that require `unsafe` but don't have `unsafe` blocks around them, making it hard to figure out where `unsafe` usage is required. If we can, we should enable Rust Edition 2024, which should help with linting for these kinds of issues. It's early enough in the C# client's development that I think we can reasonably enable this now.